### PR TITLE
feat(MLOP-1985): optional params

### PR DESCRIPTION
**Pull request by ralphrass**
## Why? :open_book:
The `eager_evaluation` param forces Spark to apply the currently mapped changes to a DataFrame. When this parameter is set to False, Spark follows its standard behavior of lazy evaluation. Lazy evaluation can improve Spark's performance as it allows Spark to build the best version of the execution plan.
The `row_deduplication` is not applicable to all scenarios. For instance, when using Delta operations, row deduplication can be done through `merge` operations.

## What? :wrench:
- Add new arguments to enable/disable eager evaluation and row deduplication in Source and FeatureSet classes.

## Type of change
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Release

## How everything was tested? :straight_ruler:
- Databricks notebooks

### CODE REVIEW

File: butterfree/extract/source.py [hunk 1]
```diff
@@ -49,13 +49,22 @@ class Source(HookableComponent):
...
+        mapped changes to the DataFrame. When this parameter is set to
+        False, Spark follows its standard behaviour of lazy evaluation.
+        Lazy evaluation can improve Spark's performance as it allows
+        Spark to build the best version of the execution plan.
+
     """
 
-    def __init__(self, readers: List[Reader], query: str) -> None:
+    def __init__(
+        self, readers: List[Reader], query: str = "", eager_evaluation: bool = True,
```
Comment (michellyrds) [inline butterfree/extract/source.py:61]: why did you add this empty string as default value?

**Review by michellyrds — APPROVED** [review]

File: butterfree/extract/source.py (hunk 1 above)
Comment (albjoaov) [reply butterfree/extract/source.py:61]: 👀

Comment (albjoaov) [inline butterfree/extract/source.py:61]: By the docs, we have: "Lazy evaluation can improve Spark's performance as it allows Spark to build the best version of the execution plan.". Looks like that `False` as default is better, or am I wrong?

Comment (albjoaov) [reply butterfree/extract/source.py:61]: Or is it `True` to maintain some compatibility?

**Review by albjoaov — APPROVED** [review]

File: butterfree/extract/source.py (hunk 1 above)
Comment (ralphrass) [reply butterfree/extract/source.py:61]: You're right, it's defaulted `True` so we can turn it off gradually in our pipelines.

Comment (ralphrass) [reply butterfree/extract/source.py]: Fixed

**Review by michellyrds — APPROVED** [review]
